src/function/table/read_file.cpp

```diff
@@ -0,0 +1,229 @@
+static void ReadFileExecute(ClientContext &context, TableFunctionInput &input, DataChunk &output) {
+	auto &bind_data = input.bind_data->Cast<ReadFileBindData>();
+	auto &state = input.global_state->Cast<ReadFileGlobalState>();
+	auto &fs = FileSystem::GetFileSystem(context);
+
+	auto output_count = MinValue<idx_t>(STANDARD_VECTOR_SIZE, bind_data.files.size() - state.current_file_idx);
+
+	// We utilize projection pushdown here to only read the file content if the 'data' column is requested
+	for (idx_t out_idx = 0; out_idx < output_count; out_idx++) {
+		// Add the file name to the output
+		auto &file_name = bind_data.files[state.current_file_idx + out_idx];
+
+		unique_ptr<FileHandle> file_handle = nullptr;
+
+		// Given the columns requested, do we even need to open the file?
+		if (state.requires_file_open) {
+			file_handle = fs.OpenFile(file_name, FileFlags::FILE_FLAGS_READ);
+		}
+
+		for (idx_t col_idx = 0; col_idx < state.column_ids.size(); col_idx++) {
+			// We utilize projection pushdown to avoid potentially expensive fs operations.
+			auto proj_idx = state.column_ids[col_idx];
+			try {
+				switch (proj_idx) {
+				case ReadFileBindData::FILE_NAME_COLUMN: {
+					auto &file_name_vector = output.data[col_idx];
+					auto file_name_string = StringVector::AddString(file_name_vector, file_name);
+					FlatVector::GetData<string_t>(file_name_vector)[out_idx] = file_name_string;
+				} break;
+				case ReadFileBindData::FILE_CONTENT_COLUMN: {
+					auto file_size = file_handle->GetFileSize();
+					AssertMaxFileSize(file_name, file_size);
+					auto &file_content_vector = output.data[col_idx];
+					auto content_string = StringVector::EmptyString(file_content_vector, file_size);
+					file_handle->Read(content_string.GetDataWriteable(), file_size);
+					content_string.Finalize();
+
+					OP::VERIFY(file_name, content_string);
+
+					FlatVector::GetData<string_t>(file_content_vector)[out_idx] = content_string;
+				} break;
+				case ReadFileBindData::FILE_SIZE_COLUMN: {
+					auto &file_size_vector = output.data[col_idx];
+					FlatVector::GetData<int64_t>(file_size_vector)[out_idx] = file_handle->GetFileSize();
+				} break;
+				case ReadFileBindData::FILE_LAST_MODIFIED_COLUMN: {
+					auto &last_modified_vector = output.data[col_idx];
+					// This can sometimes fail (e.g. httpfs file system cant always parse the last modified time
+					// correctly)
+					try {
+						auto timestamp_seconds = Timestamp::FromEpochSeconds(fs.GetLastModifiedTime(*file_handle));
+						FlatVector::GetData<timestamp_t>(last_modified_vector)[out_idx] = timestamp_seconds;
+					} catch (ConversionException &) {
+						FlatVector::SetNull(last_modified_vector, out_idx, true);
+					}
+				} break;
+				default:
+					FlatVector::SetNull(output.data[col_idx], out_idx, true);
+				}
+			}
+			// Filesystems are not required to support all operations, so we just set the column to NULL if not
+			// implemented
+			catch (NotImplementedException &) {
```

Comment:
This should use the new `ErrorData` to correctly deal with RTTI in extensions in all cases